File: tools/ocaml/xenstored/select.ml
(*
 * Copyright (C) 2014 Zheng Li <dev@zheng.li>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; version 2.1 only. with the special
 * exception on linking described in file LICENSE.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *)


(* The [read], [write], [except] are fields mapped to the POLLIN/OUT/PRI
   subscription flags used by poll, which have a correspondence to the
   readfds, writefds, exceptfds concept as in select. *)
type event = {
	mutable read: bool;
	mutable write: bool;
	mutable except: bool;
}

external select_on_poll: (Unix.file_descr * event) array -> int -> int = "stub_select_on_poll"

let init_event () = {read = false; write = false; except = false}

let select in_fds out_fds exc_fds timeout =
	let h = Hashtbl.create 57 in
	let add_event event_set fd =
		let e =
			try Hashtbl.find h fd
			with Not_found ->
				let e = init_event () in
				Hashtbl.add h fd e; e in
		event_set e in
	List.iter (add_event (fun x -> x.read <- true)) in_fds;
	List.iter (add_event (fun x -> x.write <- true)) out_fds;
	List.iter (add_event (fun x -> x.except <- true)) exc_fds;
	(* Unix.stdin and init_event are dummy input as stubs, which will
           always be overwritten later on.  *)
	let a = Array.make (Hashtbl.length h) (Unix.stdin, init_event ()) in
	let i = ref (-1) in
	Hashtbl.iter (fun fd event -> incr i; Array.set a !i (fd, event)) h;
	let n = select_on_poll a (int_of_float (timeout *. 1000.)) in
	let r = [], [], [] in
	if n = 0 then r else
		Array.fold_right
			(fun (fd, event) (r, w, x) ->
			 (if event.read then fd :: r else r),
			 (if event.write then fd :: w else w),
			 (if event.except then fd :: x else x))
			a r

File: tools/vnet/vnetd/select.c
/*
 * Copyright (C) 2003 - 2004 Mike Wray <mike.wray@hp.com>.
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of the
 * License, or  (at your option) any later version. This library is 
 * distributed in the  hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 */

#include <stdlib.h>
#include <errno.h>
#include <unistd.h>

#include "select.h"

/** Zero all the file descriptor sets.
 *
 * @param set select set
 * @param fd file descriptor
 * @return 0 on success, -1 otherwise
 */
void SelectSet_zero(SelectSet *set){
    set->n = 0;
    FD_ZERO(&set->rd);
    FD_ZERO(&set->wr);
    FD_ZERO(&set->er);
}

/** Add a file descriptor to the write set.
 *
 * @param set select set
 * @param fd file descriptor
 * @return 0 on success, -1 otherwise
 */
void SelectSet_add_read(SelectSet *set, int fd){
    FD_SET(fd, &set->rd);
    if(fd > set->n) set->n = fd;
}

/** Add a file descriptor to the write set.
 *
 * @param set select set
 * @param fd file descriptor
 * @return 0 on success, -1 otherwise
 */
void SelectSet_add_write(SelectSet *set, int fd){
    FD_SET(fd, &set->wr);
    if(fd > set->n) set->n = fd;
}

/** Select on file descriptors.
 *
 * @param set select set
 * @param timeout timeout (may be NULL for no timeout)
 * @return 0 on success, -1 otherwise
 */
int SelectSet_select(SelectSet *set, struct timeval *timeout){
    return select(set->n+1, &set->rd, &set->wr, &set->er, timeout);
}
